### PR TITLE
Add IRQ-driven buffered USART implementation for STM32 v2 usart

### PR DESCRIPTION
**Pull request by lulf**
* Implementation based on nRF UARTE, but simplified to not use DMA to
  avoid complex interaction between DMA and USART.
* Implementation of AsyncBufRead and AsyncWrite traits
* Some unit tests to ring buffer

### CODE REVIEW

**Comment by lulf** [comment]
Needs https://github.com/embassy-rs/stm32-data/pull/75

File: embassy-stm32/src/usart/v2.rs [hunk 1]
```diff
@@ -153,25 +159,224 @@ impl<'d, T: Instance, RxDma> embedded_hal::blocking::serial::Write<u8>
...
+
+            state.rx_waker.register(cx.waker());
+            Poll::<Result<&[u8], embassy::io::Error>>::Pending
+        })
+    }
+    fn consume(mut self: Pin<&mut Self>, amt: usize) {
+        self.inner.with(|state| {
+            state.rx.pop(amt);
+        });
+        self.inner.pend();
```
Comment (timokroeger) [inline embassy-stm32/src/usart/v2.rs:356]: Would it be enough to call pend only when the buffer was full?

File: embassy-stm32/src/usart/v2.rs [hunk 2]
```diff
@@ -153,25 +159,224 @@ impl<'d, T: Instance, RxDma> embedded_hal::blocking::serial::Write<u8>
...
+                return Poll::Pending;
+            }
+
+            let n = core::cmp::min(tx_buf.len(), buf.len());
+            tx_buf[..n].copy_from_slice(&buf[..n]);
+            state.tx.push(n);
+
+            Poll::Ready(Ok(n))
+        });
+        self.inner.pend();
```
Comment (timokroeger) [inline embassy-stm32/src/usart/v2.rs:379]: I think it is only required to manually pend the ISR when the buffer was empty. If there was data still in the buffer the TXE interrupt should still be enabled.

File: examples/stm32l0/src/bin/usart_irq.rs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            interrupt::take!(USART1),
+            &mut TX_BUFFER,
+            &mut RX_BUFFER,
+        )
+    };
+
+    usart.write_all(b"Hello Embassy World!\r\n").await.unwrap();
+    info!("wrote Hello, starting echo");
+
+    let mut buf = [0; 1];
```
Comment (timokroeger) [inline examples/stm32l0/src/bin/usart_irq.rs:46]: Maybe choose a larger buffer size e.g. 4, to illustrate that the buffered UART can efficiently handle multibyte receives and might read *up to* 4 bytes.

**Review by timokroeger — COMMENTED** [review]
Looks good overall, thank you for the efforts!
I did a quick port to USARTv1 for testing (because I do not have v2 hardware) and it worked without any major changes.

Once it is merged we can think about moving the buffered UART code to a shared place where we can have it for both v1 and v2.

Furthermore I think there is the potential to reduce the interrupt overhead even further if we enable and check IRQs only as required. E.g only run `on_tx()` when the TXIE bit is set.

**Comment by lulf** [comment]
@timokroeger Thanks for the review. I have address your comments. 

PR now points to stm32-data PR https://github.com/embassy-rs/stm32-data/pull/78 to fix an issue with the USART IRQ pins

File: embassy-hal-common/src/ring_buffer.rs [hunk 4]
```diff
@@ -67,6 +67,14 @@ impl<'a> RingBuffer<'a> {
         self.empty = self.start == self.end;
     }
 
+    pub fn is_full(&mut self) -> bool {
```
Comment (bobmcwhirter) [inline embassy-hal-common/src/ring_buffer.rs:70]: Probably not important, but `&mut` isn't actually required is it?

**Review by bobmcwhirter — APPROVED** [review]
LGTM

**Pull request merged**